jwt-post.md

```diff
@@ -0,0 +1,171 @@
+
+```
+<% String eid = request.getParameter("eid"); %>
+	...
+	Employee ID: <%= eid %>
+```
+
+Imagine the nightmares you'll have after clicking http://example.com/employee.jsp?eid=alert%28%22you%20have%20been%20p0wned%22%29...
+
+## Cookie or header?
+
+If you put your JWTs in a cookie, you need to take precautions to combat CSRF.
+If you use secure, HTTP-only cookies, you don't need to worry about XSS, however, because scripts don't have access to the content of such cookies.
+There's no way someone can abuse XSS and take your JWT to impersonate you.
+
+If you put your JWTs in a header, you don't need to worry about CSRF.
+You do need to worry about XSS, however.
+If someone can abuse XSS to steal your JWT, this person is able to impersonate you.
+
+In my 2016 presentation, I stated that "defense against CSRF is straightforward and durable."
+This statement was based on advice offered by the [Open Web Application Security Project (OWASP)](https://www.owasp.org/) at that time.
+Years later, defense against CSRF is still durable, but a little less straightforward.
+We'll come back to that in a minute.
+
+XSS, on the other hand, is something you need to constantly keep in mind.
+Each template you add could open up possibilities for XSS.
+The same holds for all those NPM packages you add to your front-end project, either directly or indirectly.
+
+My conclusion from this is that JWTs belong in a secure, HTTP-only cookie, and should be used in combination with preventive measures against CSRF.
+
+## Seeing is believing
+
+Because the proof of the pudding is in the eating, I wrote a simple front-end app and two back-end apps that demonstrate a session-based and JWT-based approach to authentication: https://github.com/ljpengelen/java-meetup-jwt
+
+With a simple `docker-compose` command, you can start three instances of either of the two back ends, a database, and an instance of nginx that serves the front end and acts as load balancer.
+You can open the front end in your browser, create an account, log in, and then stop some of the back-end instances with `docker stop`.
+
+In the case of the JWT-based back end, it doesn't matter which two instances you stop.
+In the case of the session-based back end, stopping the instance your connected to will terminate your session.
+
+# Measures against CSRF
+
+The OWASP has a [cheat sheet about measures against CSRF](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html).
+The applications mentioned above use two of those measures.
+First, they combat CSRF by checking the `Origin` and `Referer` headers.
+Second, they require that each request contains a secure random token, generated by a cryptographically secure random-number generator.
+Keeping track of the CSRF tokens in the front end is not completely straightforward, but it's an acceptable price to pay if you ask me.
```

Comment:
Maybe you can elaborate on this a little? As I reader I was left wondering what kind of complexity we're talking about here.